## Fonksiyonlar

#### Neden İhtiyacımız Var?

1. Yazacağımız kodları alt parçalara bölebilmek ve kodun anlaşılabilirliğini arttırmak için
2. Program içerisinde birden çok kez tekrar edicek durumları tek bir fonksiyon içerisinde yazıp kod tekrarını önlemek için

#### Fonksiyon Tanımlanması

###### Syntax Yapısı:

def fonksiyon_ismi(fonksiyon_argumanlari):

    fonksiyonda_yapilacak_islemler
    
#### ÖNEMLİ!
Her fonksiyon bir argüman almak zorunda değildir, eğer ki argüman almıyorsa parantez içerisi boş bırakılır. Fonksiyonun parametre alıp almayacağı fonksiyonun yapacağı işe bağlıdır.

#### Parametresiz Fonksiyon ve Fonksiyon Çağrısı
Eğer ki fonksiyon bir parametre almayacaksa, direkt olarak fonksiyon_ismi() şeklinde çağrılabilir

In [4]:
def selamla():
    print("Merhaba")
selamla()

Merhaba


#### Parametreli Fonksiyon ve Fonksiyon Çağrısı
Eğer ki fonksiyon bir parametre alacaksa, fonksiyon_ismi(arguman) şeklinde çağrılabilir

In [8]:
def selamla2(isim):
    print(type(isim))
    print("Merhaba",isim)
    
selamla2("Ogun")
selamla2("Oguz")
selamla2(5)

<class 'str'>
Merhaba Ogun
<class 'str'>
Merhaba Oguz
<class 'int'>
Merhaba 5


#### Varsayılan Parametre
Bazı durumlarda dışarıdan alacağımız parametrelere varsayılan değerler atamamız gerekebilir. Eğer ki parametreli tanımlanmış fonksiyona çağrılma esnasında bir parametre gönderilmezse, TypeError hatası alacağız. İşte bu duruma önlem için varsayılan parametre tanımlanır, ve parametre gönderilmediği durumda varsayılan parametre kullanılır.

In [11]:
def selamla(isim="Oguz"):
    print("Merhaba",isim)
selamla("Ogun")
selamla()

Merhaba Ogun
Merhaba Oguz


##### ÖNEMLİ

Eğer ki fonksiyon birden fazla parametre alıyorsa ve ilk parametre varsayılan değere sahipse, ondan sonraki parametre de varsayılan değere sahip olmak zorundadır

In [16]:
def selamla2(isim="Ogun",isim2="Oguz"):
    print("Merhaba {} ve {}".format(isim,isim2))
selamla2("Ogun","Oguz")
selamla2("Osman","Neriman")

Merhaba Ogun ve Oguz
Merhaba Osman ve Neriman


#### Parametre İsmi ile Fonksiyon Çağrısı

Fonksiyon çağırılırken eğer ki parametre ismi verirsek parametrelerin sırasının bir önemi yoktur.

In [17]:
selamla2(isim2="Neriman",isim="Osman")

Merhaba Osman ve Neriman


#### Değişken Sayıda Parametre

Fonksiyon tanımlanması esnasında kaç adet parametre tanımlandıysa, fonksiyon çağrısı esnasında da o kadar sayıda parametre vermemiz gerekir. Ancak bazı durumlarda fonksiyonun alacağı parametre sayısı değişkenlik gösterebilir.

In [21]:
def elemanYazdir(*sayilar):
    for sayi in sayilar:
        print(sayi,end=" ")
    print(end="\n")
    
elemanYazdir(1,12,3,84,54,121)
elemanYazdir(5)
elemanYazdir(12,54,36)
elemanYazdir()

1 12 3 84 54 121 
5 
12 54 36 



#### Fonksiyonların Geriye Sonuç Döndürmesi
Fonksiyonlar, işlem yapıldıktan sonra geriye bir veya birden fazla sonuç döndürebilir. Döndürülen sonuçlar değişken veya değişkenlerde saklanmalıdır. Sonuç veya sonuçlar return keywordü ile geriye döndürülür, return komutu çalıştığı zaman fonksiyon sonlandırılır.

##### Fonksiyonların Geriye Tek Bir Sonuç Döndürmesi

In [26]:
def topla(sayi1,sayi2):
    return sayi1+sayi2
    
sonuc=topla(15,50)
print(sonuc)

65


##### Fonksiyonların Geriye Birden Çok Sonuç Döndürmesi

In [29]:
def listeBuyukKucuk(liste):
    liste.sort()
    return liste[-1],liste[0]
buyuk,kucuk=listeBuyukKucuk([1,7,12,3,5,96,35])
print(kucuk,buyuk)

1 96


#### Özyinelemeli Fonksiyonlar

Kendi kendini çağıran fonksiyonlara özyinelemeli fonksiyon denilir. Bazı durumlarda fonksiyon içerisinde tekrar fonksiyon çağırılması gerekebilir

In [37]:
def faktoriyelHesapla(sayi):
    if sayi==1:
        return 1
    return sayi*faktoriyelHesapla(sayi-1)
print(faktoriyelHesapla(5))

120


##### Yukarıdaki Kodun Çalışma Mantığı

İlk olarak fonksiyon 5 sayısı ile çağırılmıştır ve sayi değişkenine 5 atanır

1. Fonksiyon Çağrısı-> return 5*faktoriyelHesapla(4)         -> Kendi içerisinde faktoriyelHesapla fonksiyonu 4 ile çağırılır

2. Fonksiyon Çağrısı-> return 4*faktoriyelHesapla(3)         -> Kendi içerisinde faktoriyelHesapla fonksiyonu 3 ile çağırılır

3. Fonksiyon Çağrısı-> return 3*faktoriyelHesapla(2)         -> Kendi içerisinde faktoriyelHesapla fonksiyonu 2 ile çağırılır

4. Fonksiyon Çağrısı-> return 2*faktoriyelHesapla(1)         -> Kendi içerisinde faktoriyelHesapla fonksiyonu 1 ile çağırılır

5. Fonksiyon Çağrısı-> return 1

Yani,

faktoriyelHesapla(1) sonucu-> 1

faktoriyelHesapla(2) sonucu-> 2*faktoriyelHesapla(1)=2

faktoriyelHesapla(3) sonucu-> 3*faktoriyelHesapla(2)=6

faktoriyelHesapla(4) sonucu-> 4*faktoriyelHesapla(3)=24

faktoriyelHesapla(5) sonucu-> 5*faktoriyelHesapla(4)=120


#### Local ve Global Kavramı
Fonksiyon içerisinde tanımladığımız değişkenler local değişkenlerdir ve sadece fonksiyon blokları içerisinde kullanılabilir.
Main fonksiyon içerisinde tanımladığımız değişkenler ise global değişkenlerdir ve tüm kod blokları içerisinde kullanılabilir.

In [41]:
def sayiDegistir(x):
    print(y)
    z=int()#local degisken
    x=5#local degisken
y=50#global degisken
sayiDegistir(y)
print(y)

50
50
